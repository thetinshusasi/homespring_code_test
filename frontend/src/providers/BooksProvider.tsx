import React, { useEffect, useState } from "react";
import useGetBooksBySearchTerm from "../hooks/useGetBooksBySearchTerm";
import Book from "../models/books/Book"

export interface BooksContextProps {
    books: Book[] | undefined,
    startIndex: number,
    kind: string,
    totalItems: number,
    error: any,
    loading: boolean,
    getBooksBySearchTerm: (val: string) => void
}
interface BooksProviderProps {
    children: React.ReactNode;
}
export const BooksContext = React.createContext<BooksContextProps>({
    books: undefined,
    error: false,
    getBooksBySearchTerm: () => { },
    kind: "",
    loading: false,
    startIndex: 0,
    totalItems: 0

});

const BooksProvider = ({ children }: BooksProviderProps) => {
    const { data, error, getBooksBySearchTerm, loading } = useGetBooksBySearchTerm();
    const { items: books = undefined, startIndex = 0, kind = "", totalItems = 0 } = data || {}

    const [value, setValue] = useState({
        books,
        error,
        getBooksBySearchTerm,
        loading,
        startIndex,
        totalItems,
        kind
    })

    useEffect(() => {
        const { items: books = undefined, startIndex = 0, kind = "", totalItems = 0 } = data || {}
        setValue({
            books,
            error,
            loading,
            startIndex,
            totalItems,
            kind,
            getBooksBySearchTerm

        })


    }, [data, loading, error])


    return (
        <BooksContext.Provider
            value={value}
        >
            {children}
        </BooksContext.Provider>
    );
};

export default BooksProvider;
