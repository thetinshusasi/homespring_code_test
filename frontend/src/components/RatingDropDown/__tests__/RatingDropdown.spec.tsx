/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable jest/valid-expect */

import React from 'react';
import RatingDropdown, { IRatingDropdownProps } from '../index';
import BooksProvider, { BooksContextProps } from '../../../providers/BooksProvider';
import { fireEvent, getByTestId, render, act } from '@testing-library/react';

// has to be defined like this cannot be global

describe('CardItem Component', () => {
    let props: IRatingDropdownProps;

    beforeEach(() => {
        props = {
            ratings: [1, 2, 3, 4],
            values: 2
        };
    });

    it('>> renders component', () => {
        const contextValue: BooksContextProps = {
            books: undefined,
            error: false,
            getBooksBySearchTerm: () => { },
            kind: "",
            loading: false,
            startIndex: 0,
            totalItems: 0
        };

        const { container } = render(
            <BooksProvider value={contextValue}>
                <RatingDropdown {...props} />
            </BooksProvider>
        );
        expect(container).toMatchSnapshot();
    });

    it('>> renders component and contains all props', async () => {
        const mockFn = jest.fn()
        const contextValue: BooksContextProps = {
            books: undefined,
            error: false,
            getBooksBySearchTerm: mockFn,
            kind: "",
            loading: false,
            startIndex: 0,
            totalItems: 0
        };
        const { getByText } = render(
            <BooksProvider value={contextValue}>
                <RatingDropdown {...props} />
            </BooksProvider>
        );
        expect(getByText('Filter your results'))


    });
    it('>> renders component and pm', async () => {
        const mockFn = jest.fn()
        const contextValue: BooksContextProps = {
            books: undefined,
            error: false,
            getBooksBySearchTerm: mockFn,
            kind: "",
            loading: false,
            startIndex: 0,
            totalItems: 0
        };
        props = {
            ...props,
            onSelectChange: mockFn
        }
        const { getByText } = render(
            <BooksProvider value={contextValue}>
                <RatingDropdown {...props} />
            </BooksProvider>
        );


        fireEvent.click(getByText('Greater than 2 star'))

    });


});
