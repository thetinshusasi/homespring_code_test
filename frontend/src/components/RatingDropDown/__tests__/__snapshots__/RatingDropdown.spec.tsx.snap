// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem Component >> renders component 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiTypography-root makeStyles-heading-3 MuiTypography-h6 MuiTypography-gutterBottom"
      >
        Filter your results
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div>
        <div
          class="MuiFormControl-root makeStyles-formControl-1"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
            data-shrink="true"
            id="demo-simple-select-outlined-label"
          >
            All books
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            data-testid="ratingSelect"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="demo-simple-select-outlined-label demo-simple-select-outlined"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
              id="demo-simple-select-outlined"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Greater than 2 star
                </span>
              </div>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="2"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-6 PrivateNotchedOutline-legendNotched-7"
              >
                <span>
                  All books
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
